action_sync/src/file.rs

```diff
@@ -0,0 +1,263 @@
+        self.header()
+    }
+}
+
+impl Iterator for ActionsFileReader {
+    type Item = (Block, Vec<ContextActionMessage>);
+
+    /// Return a tuple of a block and list action in the block
+    fn next(&mut self) -> Option<Self::Item> {
+        self.cursor = match self.reader.seek(SeekFrom::Start(self.cursor)) {
+            Ok(c) => c,
+            Err(_) => {
+                return None;
+            }
+        };
+        let mut h = [0_u8; 4];
+        let _ = self.reader.read_exact(&mut h);
+        let content_len = u32::from_be_bytes(h);
+        if content_len <= 0 {
+            return None;
+        }
+        let mut b = BytesMut::with_capacity(content_len as usize);
+        unsafe { b.set_len(content_len as usize) }
+        let _ = self.reader.read_exact(&mut b);
+
+        let reader = snap::read::FrameDecoder::new(b.reader());
+
+        let item = match bincode::deserialize_from::<_, (Block, Vec<ContextActionMessage>)>(reader)
+        {
+            Ok(item) => item,
+            Err(_) => {
+                return None;
+            }
+        };
+        self.cursor += h.len() as u64 + content_len as u64;
+        Some(item)
+    }
+}
+
+/// # ActionFileWriter
+///
+/// writes block and list actions to file in `path`
+pub struct ActionsFileWriter {
+    header: ActionsFileHeader,
+    file: File,
+}
+
+impl ActionsFileWriter {
+    pub fn new<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let file = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .read(true)
+            .open(path)?;
+        let mut reader = BufReader::new(file.try_clone()?);
+        let _ = reader.seek(SeekFrom::Start(0));
+        let mut h = [0_u8; HEADER_LEN];
+        let _ = reader.read_exact(&mut h);
+        let header = ActionsFileHeader::from(h);
+        Ok(ActionsFileWriter { file, header })
+    }
+
+    pub fn header(&self) -> ActionsFileHeader {
+        self.header
+    }
+}
+
+unsafe impl Send for ActionsFileWriter {}
+
+unsafe impl Sync for ActionsFileWriter {}
+
+impl ActionsFileWriter {
+    pub fn update(&mut self, block: Block, actions: Vec<ContextActionMessage>) -> Result<u32> {
+        let block_level = block.block_level;
+        let actions_count = actions.len() as u32;
+        let block_hash = block.block_hash;
+        self._fetch_header();
+
+        // Check if currently saved block precedes the incoming block
+        if block.predecessor != self.header.current_block_hash && self.header.block_count > 0 {
+            return Err(anyhow!("Block out of sequence"));
+        }
+
+        let mut out = Vec::new();
+        let writer = snap::write::FrameEncoder::new(&mut out);
+        bincode::serialize_into(writer, &(block, actions))?;
+
+        // Writes the header if its not already set
+        if self.header.block_count <= 0 {
+            let header_bytes = self.header.to_vec();
+            let _ = self.file.seek(SeekFrom::Start(0));
```

Comment:
consider unwraping result instead of ignoring it

Reply:
alright